include/nbl/system/demote_promote_writer_readers_lock.h

```diff
@@ -1,64 +1,88 @@
+namespace impl
 {
-	constexpr static inline uint32_t MaxActors = 1023;
-	constexpr static inline uint32_t SpinsBeforeYield = 5000;
-	struct StateSemantics
+	template<typename U>
+	struct DPWRLstate_lock
+	{
+		using type = std::atomic<U>;
+	};
+
+	// will fail if `atomic_unsiged_lock_free` doesn't exist or its value type has wrong size
+	template<typename U> requires (sizeof(std::atomic_unsigned_lock_free::value_type) >= sizeof(U))
+	struct DPWRLstate_lock<U>
 	{
-		explicit inline operator uint32_t() const { return std::bit_cast<uint32_t>(*this); }
+		using type = std::atomic_unsigned_lock_free;
+	};
+
+	constexpr inline DPWRLstate_lock<uint32_t>::type::value_type DPWRLMaxActors = 1023;
 
+	struct DPWRLStateSemantics
+	{
+		using state_lock_value_t = typename DPWRLstate_lock<uint32_t>::type::value_type;
 		uint32_t currentReaders : 10 = 0;
 		uint32_t pendingWriters : 10 = 0;
 		uint32_t pendingUpgrades : 10 = 0;
 		uint32_t writing : 1 = 0;
-		uint32_t stateLocked : 1 = 0; 
+		uint32_t stateLocked : 1 = 0;
+
+		consteval inline operator state_lock_value_t() const
+		{ 
+			assert(currentReaders <= DPWRLMaxActors && pendingWriters <= DPWRLMaxActors && pendingUpgrades <= DPWRLMaxActors);
+			return static_cast<state_lock_value_t>((currentReaders << 22) | (pendingWriters << 12) | (pendingUpgrades << 2) | (writing << 1) | stateLocked);
+		}
 	};
-	boost::atomic_unsigned_lock_free state;
+} //namespace impl
 
-	constexpr uint32_t flipLock = StateSemantics{ .currentReaders = 0,.pendingWriters = 0,.pendingUpgrades = 0,.writing = false,.stateLocked = true };
-	constexpr uint32_t writingMask = StateSemantics{ .currentReaders = 0,.pendingWriters = 0,.pendingUpgrades = 0,.writing = true,.stateLocked = false };
+class demote_promote_writer_readers_lock
+{
+public:
+	using state_lock_t = impl::DPWRLstate_lock<uint32_t>::type;
+	using state_lock_value_t = impl::DPWRLStateSemantics::state_lock_value_t;
+	// Limit on how many threads can be launched concurrently that try to read/write from/to the same file
+	constexpr static inline state_lock_value_t MaxActors = impl::DPWRLMaxActors;
+	
+	friend class dp_read_lock_guard;
+	friend class dp_write_lock_guard;
 
 	/**
-	* @brief Acquires lock for reading. This is the operation with the lowest priority, thread will be blocked until there are no writers and no readers waiting for a writer upgrade
+	* @brief Acquires lock for reading. This thread will be blocked until there are no writers (writing or pending) and no readers waiting for a writer upgrade
 	*/
 	void read_lock()
 	{
-		constexpr uint32_t preemptedMask = StateSemantics{ .currentReaders = 0,.pendingWriters = MaxActors,.pendingUpgrades = MaxActors,.writing = true,.stateLocked = false };
-		for (uint32_t spinCount = 0; true; spinCount++)
+		constexpr state_lock_value_t preemptedMask = pendingWritersMask | pendingUpgradesMask | writingMask;
+		for (state_lock_value_t spinCount = 0; true; spinCount++)
```

Comment:
you don't want to use `state_lock_value_t` for this, it has nothing in common/todo with it